Add support for showing custom resources in the dashboard

davidfowl opened the pull request:
- Added DashboardAnnotation which allows changing a dashboard state. There are 2 methods, one to create the initial state from resource annotations and another to send updates to the dashboard. These are immutable snapshots to avoid thread safety issues.
- Show all resources based on the app model before watching dcp for updates.


https://github.com/dotnet/aspire/assets/95136/487352a7-ac73-4a1d-8e5f-0194341daae0



Left TODO:
- ~Replicas are broken on the resources view with this change. We need to delete the existing project resources if there are replicas.~
- ~Tests?~

Follow up work to enable logging.

Fixes #436 
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2390)

davidfowl committed:
Add support for showing custom resources in the dashboard
- Added DashboardAnnotation which allows changing a dashboard context. This context consists of properties, urls and the resource state.
- Show all resources based on the app model before watching dcp for updates.

Added a model for upates
- Use a snapshotting model for dashboard updates

Update the previous state

Clean up how we resolve projects from DCP resources

Some feedback
- Make the initial snapshot state take computed values from well known annotations.

Fixed typos

Small change

Added a public DashboardKnownProperties

Added playground for testing custom resource API

Remove manifest

Handle missing parameter values

Fix configuration

Apply feedback and made more changes
- Clean up te threading model. There can only be a single producer and consumer. The producer is responsible to keeping the state around if small mutations are to be made.
- Keep the resource type filter working

Remove exit code, fix create timestamp

ExecutionContext not used

Flow the cancellation token

Move to single loop

Remove logs hack for now

davidfowl commented on src/Aspire.Hosting/Dashboard/GenericResourceSnapshot.cs:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    // TODO: Support logs for generic resources.
+    internal static async IAsyncEnumerable<IReadOnlyList<(string Content, bool IsErrorMessage)>> GetLogsEnumerable()
```
Follow up PRs will enable console logs and we can decide if we want to support OTLP as well.

davidfowl commented on src/Aspire.Hosting/Dashboard/GenericResourceSnapshot.cs:
```diff
@@ -0,0 +1,36 @@
+        if (resource is IResourceWithConnectionString connectionString)
+        {
+            var cs = connectionString.GetConnectionString();
+            yield return ("ConnectionString", cs is null ? Value.ForString(cs) : Value.ForNull());
```
We need to be able to mark values as secret @drewnoakes

drewnoakes replied:
It should be straightforward to extend the protocol for this. You can add it to the objects involved in this layer now if you like, and we can plumb that through later.

We could differentiate between secrets that require permissions and secrets that just require making in the UI by default.

davidfowl replied:
Filed https://github.com/dotnet/aspire/issues/2415

davidfowl commented on src/Aspire.Hosting/Dashboard/DashboardResoruceExtensions.cs:
```diff
@@ -0,0 +1,25 @@
+    /// <param name="builder">The resource builder.</param>
+    /// <param name="callback">The callback to customize the <see cref="DashboardResourceContext"/>.</param>
+    /// <returns>The resource builder.</returns>
+    public static IResourceBuilder<TResource> WithDashboardContext<TResource>(this IResourceBuilder<TResource> builder, Action<DashboardResourceContext> callback)
```
Not 100% sure if this is the best pattern but it needs to be deferred.

drewnoakes commented on src/Aspire.Hosting/Dashboard/DashboardAnnotation.cs:
```diff
@@ -0,0 +1,52 @@
+    /// <summary>
+    /// Represents the state of the resource.
+    /// </summary>
+    public string? State { get; set; }
```
Is this expected to change over time at all? If so, what would trigger updates? If not, I'm curious what value this adds as a static item, other than populating the column. Are there some examples?

davidfowl replied:
I moved to a model where these are immutable snapshots.

drewnoakes reviewed (commented):
Looks good. Left some comments from my phone while waiting for a plane, so apologies for the lack of depth here.

davidfowl commented on src/Aspire.Dashboard/Components/Pages/Resources.razor.cs:
```diff
@@ -39,7 +39,7 @@ public Resources()
     }
 
-    private bool Filter(ResourceViewModel resource) => _visibleResourceTypes.Contains(resource.ResourceType) && (_filter.Length == 0 || resource.MatchesFilter(_filter));
+    private bool Filter(ResourceViewModel resource) => (_filter.Length == 0 || resource.MatchesFilter(_filter));
```
Ahh good point.

davidfowl commented on src/Aspire.Hosting/Dashboard/DashboardAnnotation.cs:
```diff
@@ -0,0 +1,52 @@
+    {
+        get
+        {
+            return Properties.TryGetValue("Source", out var value) ? value : null;
```
I changed this logic, but I will do that.

davidfowl replied:
Done. We'll need to expose a public API as well

davidfowl commented on src/Aspire.Dashboard/Components/ResourcesGridColumns/SourceColumnDisplay.razor:
```diff
@@ -38,9 +38,13 @@ else if (Resource.TryGetContainerImage(out var containerImage))
                ToolTip="@containerImage" />
 }
-else
+else if (Resource.Properties.TryGetValue("Source", out var value))
```
Done

davidfowl commented on playground/CustomResources/CustomResources.AppHost/TestResource.cs:
```diff
@@ -0,0 +1,73 @@
+        {
+            if (item.TryGetLastAnnotation<DashboardAnnotation>(out var annotation))
+            {
+                var states = new[] { "Starting", "Running", "Finished" };
```
This sample brings up some interesting questions. Should it be possible to wait for the dashboard state of a resource to be initialized? I think yes

mitchdenny replied:
I'm liking this model overall. I think we should be more generic in our terminology here, I can see this annotation turning into a bit of a proxy for not only state, but also commanding against specific resources.

davidfowl commented on src/Aspire.Hosting/Extensions/ParameterResourceBuilderExtensions.cs:
```diff
@@ -37,6 +38,15 @@ public static IResourceBuilder<ParameterResource> AddParameter(this IDistributed
+                      {
+                          ResourceType = "Parameter",
+                          Properties = [
+                            ("Value", callback()),
```
This could be problematic, because we don't want this to throw because the value is missing. These callbacks are evaluated out of order with when DCP starts.

mitchdenny commented on src/Aspire.Hosting/Dashboard/DashboardAnnotation.cs:
```diff
@@ -0,0 +1,93 @@
+    /// <summary>
+    /// The exit code of the resource.
+    /// </summary>
+    public int? ExitCode { get; init; }
```
We should figure out of this is necessary on all resources.

davidfowl replied:
I'll remove it from here.

mitchdenny commented on src/Aspire.Hosting/Dashboard/DcpDataSource.cs:
```diff
@@ -104,6 +129,160 @@ await foreach (var (eventType, resource) in _kubernetesService.WatchAsync<T>(can
+        ResourceSnapshot? snapshot = default;
+
+        // If the resource is a redirect, we want to create a snapshot for the resource it redirects to.
+        if (resource.TryGetLastAnnotation<ConnectionStringRedirectAnnotation>(out var redirectAnnotation))
```
We should change the name of this annotation if we are going to use it in this way ... its scope is now much broader than connection strings.

mitchdenny commented on src/Aspire.Hosting/Dashboard/DashboardResourceExtensions.cs:
```diff
@@ -0,0 +1,25 @@
+    public static IResourceBuilder<TResource> WithDashboardState<TResource>(this IResourceBuilder<TResource> builder, Func<DashboardResourceState, DashboardResourceState> initialState)
+        where TResource : IResource
+    {
+        return builder.WithAnnotation(new DashboardAnnotation(initialState));
```
This should replace existing annotations.

JamesNK commented on playground/CustomResources/CustomResources.AppHost/TestResource.cs:
```diff
@@ -0,0 +1,73 @@
+        builder.Services.AddLifecycleHook<TestResourceLifecycleHook>();
+
+        var rb = builder.AddResource(new TestResource(name))
+                      .WithDashboardState(state => state with
```
I don't like dashboard in the name. It isn't dashboard state, it's state that happens to be shown in the dashboard.

JamesNK replied:
I guess custom resources like this. are local development only.

davidfowl replied:
Right, @mitchdenny had the same feedback though. I'm open to suggestions!

JamesNK commented on playground/CustomResources/CustomResources.AppHost/TestResource.cs:
```diff
@@ -0,0 +1,73 @@
+
+internal sealed class TestResourceLifecycleHook : IDistributedApplicationLifecycleHook, IAsyncDisposable
+{
+    private readonly CancellationTokenSource _tokenSource = new();
```
Needed because the BeforeStartAsync token is no longer canceled once the task completes?

davidfowl replied:
Right.

JamesNK commented on src/Aspire.Hosting/Dashboard/ConsoleLogPublisher.cs:
```diff
@@ -19,6 +19,7 @@ internal IAsyncEnumerable<IReadOnlyList<(string Content, bool IsErrorMessage)>>?
         {
             ExecutableSnapshot executable => SubscribeExecutable(executable),
             ContainerSnapshot container => SubscribeContainer(container),
+            GenericResourceSnapshot => GenericResourceSnapshot.GetLogsEnumerable(),
```
FYI this code is about to change. @karolz-ms is updating it to get logs from DCP. Need to think about how custom resources work in that world.

davidfowl replied:
Should work the same. I have a hack here https://github.com/dotnet/aspire/blob/321d0a20690c5e6a9cfda8558044b26769bc5e36/src/Aspire.Hosting/Dashboard/GenericResourceSnapshot.cs#L26-L40. This would OOM eventually though so would need a circular buffer to cache old results.

JamesNK commented on src/Aspire.Hosting/Dashboard/DashboardAnnotation.cs:
```diff
@@ -0,0 +1,93 @@
+/// <summary>
+/// The context for a all of the properties and URLs that should show up in the dashboard for a resource.
+/// </summary>
+public record DashboardResourceState
```
This should be a class.

JamesNK commented on src/Aspire.Hosting/Dashboard/DcpDataSource.cs:
```diff
@@ -36,22 +37,46 @@ internal sealed class DcpDataSource
+            finally
+            {
+                // Start watching resources after we have loaded the initial state
+                initialTcs.TrySetResult();
```
Why use a TCS instead of getting the task returned from Task.Run and awaiting that?

davidfowl committed:
Fix replicas by removing the placeholder resource

davidfowl committed:
Removed unnecesary changes

davidfowl committed:
Print no logs available.

davidfowl committed:
dummy -> placeholder

davidfowl committed:
Tests for the new API

mitchdenny commented on playground/CustomResources/CustomResources.AppHost/TestResource.cs:
```diff
@@ -0,0 +1,85 @@
+            {
+                var states = new[] { "Starting", "Running", "Finished" };
+
+                Task.Run(async () =>
```
`Task.Run` for every resource? I know its just for sample purposes but ....😅😅

karolz-ms reviewed (approved):
LGTM, just a couple of suggestions

mitchdenny reviewed (approved):
LGTM. Some reservations about the annotation name. But overall this is a good change (and much needed).

davidfowl committed:
Update playground/CustomResources/CustomResources.AppHost/TestResource.cs

Co-authored-by: Karol Zadora-Przylecki <karolz@microsoft.com>

JamesNK commented on src/Aspire.Dashboard/Components/Pages/Resources.razor.cs:
```diff
@@ -97,17 +96,29 @@ protected override void OnInitialized()
                 Debug.Assert(added, "Should not receive duplicate resources in initial snapshot data.");
             }
 
+            await InvokeAsync(() =>
```
Is this necessary? This is called directly from OnInitializedAsync so it should already be on the UI thread.

davidfowl replied:
Works, and I can make this sync again.

JamesNK commented on src/Aspire.Dashboard/Components/Pages/Resources.razor.cs:
```diff
@@ -97,17 +96,29 @@ protected override void OnInitialized()
+                    _visibleResourceTypes.Add(resourceType);
+                }
+
+                StateHasChanged();
```
Doesn't seem necessary

JamesNK commented on src/Aspire.Dashboard/Components/Pages/Resources.razor.cs:
```diff
@@ -116,14 +127,21 @@ await foreach (var (changeType, resource) in subscription.WithCancellation(_watc
-                    await InvokeAsync(StateHasChanged);
+                    await InvokeAsync(() =>
+                    {
+                        _visibleResourceTypes.Add(resource.ResourceType);
```
Does the resource type need to be added inside InvokeAsync?

The change will come from another thread so the hashset should change to a concurrentdictionary (key and value can be the same). Or an immutable collection.

davidfowl replied:
_visibleResourceTypes isn't thread safe

JamesNK commented on src/Aspire.Hosting/Dashboard/ConsoleLogPublisher.cs:
```diff
@@ -76,4 +78,11 @@ LogsEnumerable SubscribeContainerResource(ContainerSnapshot container)
+
+    private static async LogsEnumerable SubscribeGenericResource()
+    {
+        await Task.Yield();
```
Is this needed? Remove or add a comment saying why

davidfowl replied:
Our analyzers, I guess I could suppress the warning.

JamesNK replied:
Leaving it with a comment is fine. I brought it up because I didn't know why it was needed

JamesNK commented on src/Aspire.Hosting/Dashboard/DashboardAnnotation.cs:
```diff
@@ -0,0 +1,126 @@
+/// This is a single producer, single consumer channel model for pushing updates to the dashboard.
+/// The resource server will be the only caller of WatchAsync.
+/// </summary>
+public class DashboardAnnotation(Func<DashboardResourceState> initialState) : IResourceAnnotation
```
```suggestion
public sealed class DashboardAnnotation(Func<DashboardResourceState> initialState) : IResourceAnnotation
```

JamesNK commented on src/Aspire.Hosting/Dashboard/DcpDataSource.cs:
```diff
@@ -104,6 +111,146 @@ await foreach (var (eventType, resource) in _kubernetesService.WatchAsync<T>(can
+
+            await _onResourceChanged(snapshot, ResourceSnapshotChangeType.Upsert).ConfigureAwait(false);
+        }
+        else if (resource.TryGetLastAnnotation<DashboardAnnotation>(out var dashboardAnnotation))
```
When I see this annotation used here, a name like CustomResourceAnnotation or GenericResourceAnnotation seems more appropriate.

JamesNK commented on src/Aspire.Hosting/Dashboard/DcpDataSource.cs:
```diff
@@ -104,6 +111,146 @@ await foreach (var (eventType, resource) in _kubernetesService.WatchAsync<T>(can
+        {
+            // We have a dashboard annotation, so we want to create a snapshot for the resource
+            // and update data immediately. We also want to watch for changes to the dashboard state.
+            var state = dashboardAnnotation.GetIntialState();
```
Intial -> Initial

Is this method needed? I wonder if it could just be rolled up into WatchAsync. i.e. the initial state is returned immediately from WatchAsync.

davidfowl replied:
It's needed because the producer needs to make state changes based on the initial snapshot

davidfowl committed:
PR feedback
- Rename the annotation to CustomResourceAnnotation
- Simplify the UI
- Suppress async warning

JamesNK commented on src/Aspire.Hosting/ApplicationModel/CustomResourceAnnotation.cs:
```diff
@@ -40,7 +39,7 @@ public async Task UpdateStateAsync(DashboardResourceState state)
 /// The context for a all of the properties and URLs that should show up in the dashboard for a resource.
 /// </summary>
-public record DashboardResourceState
+public record CustomResourceState
```
sealed?

JamesNK commented on src/Aspire.Hosting/Dashboard/GenericResourceSnapshot.cs:
```diff
@@ -1,11 +1,12 @@
 namespace Aspire.Hosting.Dashboard;
 
-internal class GenericResourceSnapshot(DashboardResourceState state) : ResourceSnapshot
+internal class GenericResourceSnapshot(CustomResourceState state) : ResourceSnapshot
```
nit: CustomResourceSnapshot?

davidfowl committed:
Concurrent dictionaries everywhere
- Does not remove resource types from the filter

davidfowl committed:
Rename state to snapshot

davidfowl commented:
Thanks for the review all! I will follow up with a logs PR.
